pyro/mesh/reconstruction.py

```diff
@@ -285,3 +285,112 @@ def weno(q, order):
         q_minus[i] = weno_upwind(q[i+order-1:i-order:-1], order)
 
     return q_minus, q_plus
+
+
+def ppm_reconstruction(a, myg, idir):
+    """
+    The PPM reconstruction method proceeds as follows:
+
+    1. First, we perform an estimation of the values
+    of a, in each interface x_{i+1/2}. In order to accomplish this
+    particular task a cuartic polynomial interpolation over:
+
+    |----*----|----*----|----*----|----*----|
+      a_{i-1}     a_i     a_{i+1}   a_{i+2}
+
+    After the calculation of these values, defined as a_{i+1/2},
+    we stored them in ai[i].
+
+    2. Next, we define two values for each interface: a_{L,i+1/2} and a_{R,i+1/2},
+    to the limits from the left and the right of x_{i+1/2}, respectively. Under
+    this definition, each cell defines:
+
+      |----------------------*----------------------|
+    a_{R,i-1/2}              a_i              a_{L, i+1/2}
+
+    where a_{R,i-1/2} will be stored in ar[i], and a_{L,i+1/2} in
+    al[i].
+
+
+    3. Finally, we return the ar[:] and al[:] np.ndarray types.
+    """
+
+    ai = myg.scratch_array()
+    al = myg.scratch_array()
+    ar = myg.scratch_array()
+    dap = myg.scratch_array()
+    dam = myg.scratch_array()
+    davl = myg.scratch_array()
+    sign = myg.scratch_array()
+
+    if idir == 1:
+
+        dam.v(buf=3)[:, :] = a.v(buf=3) - a.ip(-1, buf=3)
+        dap.v(buf=3)[:, :] = a.ip(1, buf=3) - a.v(buf=3)
+
+        #Now, we should apply the Van Leer limiter for the slopes.
+        davl.v(buf=3)[:, :] = 0.5*(a.ip(1, buf=3) - a.ip(-1, buf=3))
+        sign.v(buf=3)[:, :] = np.sign(davl.v(buf=3))
+
+        davl.v(buf=3)[:, :] = np.minimum(np.abs(davl.v(buf=3)), 2.0 * np.abs(dam.v(buf=3)))
+        davl.v(buf=3)[:, :] = np.minimum(davl.v(buf=3), 2.0 * np.abs(dap.v(buf=3))) * sign.v(buf=3)
+
+        #Now, we are in the position of writing the reconstruction interpolation
+        #of the mid-zones.
+        ai.v(buf=3)[:, :] = a.v(buf=3) + 0.5*(dap.v(buf=3)) \
+                               - (davl.ip(1, buf=3) - davl.v(buf=3))/6.0
+
+        #From here we set ar and al to take the same value of ai. The values of ar and
+        #al, will be reformated accordingly to the neigbours behaviour.
+        al.v(buf=3)[:, :] = ai.v(buf=3)
```

Comment:
right now, ai is a_{i+1/2}, so this should be set to a_{i+1,L} and a_{i,R}, but you are setting it to L and R for a_i (I think).

Reply:
also, you don't have the information in 3 ghost cells at this point, because you are shifting your view.  I think you are only at 2 ghost cells here.